### PR TITLE
Local LSP server that improves functionality like item completion

### LINKED ISSUE #95: Autocompletion/syntax checking extremely slow/unusable
Autocompletion and syntax checking is extremely slow, if not unusable.

Issue has been reported on the community forum as well, known hardware/software:
openHABian and in my case openHAB snapshot #1272.

https://community.openhab.org/t/vs-code-openhab-extension/30205/244
https://community.openhab.org/t/vs-code-openhab-extension/30205/248 (video)
https://community.openhab.org/t/vs-code-openhab-extension/30205/245

### PR DESCRIPTION
**Pull request by SamuelBrucksch**
Initial LSP impl with example from MS sample LSP:
https://github.com/Microsoft/vscode-extension-samples/tree/master/lsp-sample

To test:
```
npm install
npm run compile
press f5 in vscode
```

Open a new file, and then select openhab as language or save as *.items/rules/...

This will fix #95 when all the work is done.

Signed-off-by: Samuel Brucksch <sasliga@freenet.de>

### CODE REVIEW

**Comment by SamuelBrucksch** [comment]
Unfortunately i did not find an easy way to use our local LSP as proxy for the remote LSP without reimplementing the whole LSP protocol. The vscode-languageclient can not run within the server:
https://github.com/Microsoft/vscode-languageserver-node-example/issues/36

My idea was something like this:
vscode extension -> local lsp -> client in local lsp -> remote lsp
However as the languageclient can not run in the server part (see issue from above) i cant follow that idea.

Instead i now run local and remote LSP server in parallel. In this screenshot you can see diagnostics from local and remote LSP. The green underlined part is from local LSP because its all uppercase (just a sample, no real diagnostics yet) and the re underlined parts are from openhab remote LSP:
image

So we still can implement the item completions in out local LSP server, add diagnostics to the one we ge from the remote server and also do all the other stuff in our local LSP like jumpt to reference and so on...

Its not as nice as extending the remote server, but its still better than nothing.

Maybe later when i have more time i can look into how to write an own language client, but for now this should be enough.

**Comment by MHerbst** [comment]
I will have a closer look at the implementation at the end of the week. It will also be necessary to do some modifications in the remote LSP because it currently announces to many capabilities which leads to context menu function that won't work.

**Comment by SamuelBrucksch** [comment]
Currently missing is this part:
```
       switch (event.type) {
            case 'ItemStateEvent':
                // TODO update item.documentation so that we always have latest values
                break;
           ...
    }
```
This will update the values of the documentation that is shown next to the item. It includes the current values, tags and groups it belongs to. Once that is added we have the same basic functionality as before with much faster completion.

We could then review & merge so we can add additional PRs for additional features and do other stuff inbetween without getting merge conflicts.

**Comment by SamuelBrucksch** [comment]
I think i'm currently at a stage that reflects the functionality from before but with much faster completion.

I can either continue on implementing LSP features, or wait for a review and merge and then add new PRs for new features, which would make more sense for working simultaniously on seperate features.

**Comment by SamuelBrucksch** [comment]
Maybe some remarks on what i did:
1. i seperated lsp server (server) and extension (client)
2. extension config is in root package json, {server,client}/package.json only contains dependencies and some meta info
3. slightly adapted npm scripts, now we have npm run compile that only compiles and npm watch that watches for changes. compilation is done from the project root and not for each module.
4. started a local and a remote language client to connect to both local and remote LSP server
5. Added LSP impl
6. added completion items provider that caches the items and updates them with SSE

I think thats all. Maybe we need a utils package for code that is used in both modules like IItem.ts and Item.ts. Maybe more will follow.

**Comment by SamuelBrucksch** [comment]
Looks like sign off did not work because i commited from two different PCs with different git settings. Looks like i have to do it again because 17 commits are with wrong sign off -.-

**Comment by Confectrician** [comment]
No, no need to do that.
Pullapprove is just a tool at the end, which can be a help but shouldn't prevent us from thinking.
I can see that there are sign off statements with valid data.

**Comment by Confectrician** [comment]
Also please note that this is a huge change and it will take some time for testing it.

**Comment by SamuelBrucksch** [comment]
Hey,

i have some time during the holidays to work a bit more on this or other related projects, so it would be nice if you could look over this so i can continue with other issues or features or with improving this impl.

BR

Samuel

**Comment by Confectrician** [comment]
I can't promise anything. 
It is the last working week of the year and I am still busy there.
Maybe at the weekend.

@MHerbst did you have time already?

**Comment by MHerbst** [comment]
Unfortunately not, but tomorrow is my last work day for this year. This means that I can start with testing (and reviewing) in two or three days.

**Comment by kubawolanin** [comment]
Hey @SamuelBrucksch, Thank you for your contributions. 
I've been absent from any development activities for a long time here - had to take care of my health.
Kudos to @Confectrician and @MHerbst for taking care of this project. 

> Would it be ok for you/the project, if the LSP would be written in pure JS?

I think it's ok for an initial contribution, yes. We can then think of migrating to TS later, so both projects are aligned in terms of code style.

And tests are always welcome :-) I promised myself to cover crucial parts of the code with them long ago. Sadly, didn't find time for them later on.

**Comment by SamuelBrucksch** [comment]
Hi,

thanks for your message. I now added a pure JS impl with Jest tests. Current coverage looks like this:
```
----------------------------|----------|----------|----------|----------|-------------------|
File                        |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
----------------------------|----------|----------|----------|----------|-------------------|
All files                   |    70.23 |     62.5 |    74.47 |    69.11 |                   |
 src                        |      100 |      100 |      100 |      100 |                   |
  LSPServer.js              |      100 |      100 |      100 |      100 |                   |
  Server.js                 |      100 |      100 |      100 |      100 |                   |
 src/DocumentValidation     |      100 |      100 |      100 |      100 |                   |
  DocumentValidator.js      |      100 |      100 |      100 |      100 |                   |
 src/ItemCompletion         |    51.85 |       55 |    57.14 |     52.5 |                   |
  Item.js                   |    52.63 |     12.5 |    46.67 |    52.63 |... 0,71,82,88,107 |
  ItemCompletionProvider.js |    51.61 |    65.63 |    69.23 |    52.46 |... 79,81,82,83,85 |
----------------------------|----------|----------|----------|----------|-------------------|
Jest: "global" coverage threshold for statements (90%) not met: 70.23%
Jest: "global" coverage threshold for branches (90%) not met: 62.5%
Jest: "global" coverage threshold for lines (90%) not met: 69.11%
Jest: "global" coverage threshold for functions (100%) not met: 74.47%
```

Still some tests missing. Buit overall quite good already.

**Comment by SamuelBrucksch** [comment]
I think i have all the tests in place more or less...

Unit:
```
----------------------------|----------|----------|----------|----------|-------------------|
File                        |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
----------------------------|----------|----------|----------|----------|-------------------|
All files                   |      100 |    95.74 |      100 |      100 |                   |
 src                        |      100 |      100 |      100 |      100 |                   |
  LSPServer.js              |      100 |      100 |      100 |      100 |                   |
  Server.js                 |      100 |      100 |      100 |      100 |                   |
 src/DocumentValidation     |      100 |      100 |      100 |      100 |                   |
  DocumentValidator.js      |      100 |      100 |      100 |      100 |                   |
 src/ItemCompletion         |      100 |    94.87 |      100 |      100 |                   |
  Item.js                   |      100 |     87.5 |      100 |      100 |                56 |
  ItemCompletionProvider.js |      100 |    96.77 |      100 |      100 |                31 |
----------------------------|----------|----------|----------|----------|-------------------|
```

Integration:
```
----------------------------|----------|----------|----------|----------|-------------------|
File                        |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
----------------------------|----------|----------|----------|----------|-------------------|
All files                   |     62.5 |    51.06 |    53.06 |    65.89 |                   |
 src                        |    57.45 |       25 |    27.78 |     67.5 |                   |
  LSPServer.js              |      100 |      100 |      100 |      100 |                   |
  Server.js                 |    53.49 |       25 |    27.78 |    63.89 |... 14,130,139,142 |
 src/DocumentValidation     |      100 |      100 |      100 |      100 |                   |
  DocumentValidator.js      |      100 |      100 |      100 |      100 |                   |
 src/ItemCompletion         |    63.53 |    56.41 |    66.67 |    63.53 |                   |
  Item.js                   |    52.63 |       25 |    46.67 |    52.63 |... 0,71,82,88,107 |
  ItemCompletionProvider.js |    66.67 |    64.52 |    86.67 |    66.67 |... 8,89,90,92,167 |
----------------------------|----------|----------|----------|----------|-------------------|
```

Integration could still be a bit better, for example i have one TODO still left in the code, to create, update, remove items via rest api and check if the events come in. However that is a quite complex test and i will do it later.

**Comment by SamuelBrucksch** [comment]
The config property 'openhab.restCompletions' might be obsolete now, as currently we always use rest completion. If the rest api is not available, then no items are proposed, but in all other cases we will have items completion except when 'openhab.lspEnabled' is set to false.

Is there still a need to disable rest completions?

**Comment by MHerbst** [comment]
I was able to clone the sources and create start a debug session with the new code.

@SamuelBrucksch Maybe you could also create a vsix package. That would make testing easier. I think it would also be a good idea if you could write down shortly which features are now implemented in the local LSP and should work faster then before. This PR has so many commits that I have lost track of them.

I would then concentrate on testing the extension. Javascript is not really one of my preferred languages. TypeScript is OK for me, but I don't have enough experience or knowledge to review JavaScript code.

> Is there still a need to disable rest completions? If yes we need to be able to turn it on and off in the local LSP server.

In case of installations with a large number of items, the REST calls may slow down everything. It then may help to disable the REST calls.

**Comment by SamuelBrucksch** [comment]
> I was able to clone the sources and create start a debug session with the new code.
> 
> @SamuelBrucksch Maybe you could also create a vsix package. That would make testing easier. I think it would also be a good idea if you could write down shortly which features are now implemented in the local LSP and should work faster then before. This PR has so many commits that I have lost track of them.

You can build the vsix yourself now. there is an npm script 'package' which you can run with 'npm run package'. You have to install vsce before with 'npm install vsce -g' 

Its still the list from here:
https://github.com/openhab/openhab-vscode/pull/122#issuecomment-446199834

I only added test and changed from TS to JS.

> I would then concentrate on testing the extension. Javascript is not really one of my preferred languages. TypeScript is OK for me, but I don't have enough experience or knowledge to review JavaScript code.

I tried to make the implementation clean, but maybe i did not achieve it everywhere. I used mainly JS classes so its easier to understand for Java developers. the tests and mocks look awful, i know. Its hard to understand if you never did it with jest. So maybe try to focus on the impl, that should be easy to understand. At least the logic. 

> > Is there still a need to disable rest completions? If yes we need to be able to turn it on and off in the local LSP server.
> 
> In case of installations with a large number of items, the REST calls may slow down everything. It then may help to disable the REST calls.

REST items are now cached locally, so even if REST api is slow you will not notice because for completions the local cache is used. Thats why i asked if that property still makes sense.

**Comment by SamuelBrucksch** [comment]
Here is a vsix for testing:

[openhab-0.4.1.zip](https://github.com/openhab/openhab-vscode/files/2706088/openhab-0.4.1.zip)

**Comment by SamuelBrucksch** [comment]
For me it is still ordered. However code snippets are seperate fron completion items and small and capital letters un the beginning are also seperates:

gGroup1
gGroup2
AutoHeating

Was this different before?

**Comment by MHerbst** [comment]
It is a bit strange. A problem with the sort order was first reported here: https://community.openhab.org/t/alphabetical-sorting-of-item-list-is-no-longer-possible-in-vscode-extension/61131 . I have then tried it with the version from the Marketplace and the sort order was OK. Then I installed the version you have attached here and now the sort order looks a bit strange:
image

I am not sure whether it was really introduced by your latest modifications but we need to make sure that we present the items in alphabetical order. For code completion in the editor the list is sorted correctly.

**Comment by SamuelBrucksch** [comment]
Ok thats the items list not the code completion.

Looks like this PR had some side effects noone noticed:
https://github.com/openhab/openhab-vscode/pull/114

I assumed this was only used in completion. For me the order in the items list was never ordered alphabetically. Current version from vscode marketplace gives me something like this:
image

And it already contains the above mentioned PR. So i guess the sort order should already not work then.

**Comment by SamuelBrucksch** [comment]
Indeed it does not work anymore, i opened a seperate issue:
https://github.com/openhab/openhab-vscode/issues/125

**Comment by SamuelBrucksch** [comment]
The sort order change should be a quick one, however i would propose to first merge this, so we do not get too many merge conflicts in here.
If you want the other one merged first however i can take care of it and then merge it in here or so.

**Comment by SamuelBrucksch** [comment]
Here are some open points we should still discuss so i can finish it:

1. Configuration
- should we differentiate between local and remote lsp in config?
- openhab.lspEnabled used for both or use two settings instead?
- openhab.restCompletions this is now in the local LSP and does not slow down code completion anymore, so should we keep this or remove it?

As this is a new feature, it might make sense to introduce new settings, so it is not disabled by all the people that had problems and then they will not see, that it works better now.

So for example openhab.lspEnabled is only used for remote LSP server and openhab.restCompletions is renamed to openhab.localLsp and deals with completion and other stuff. I dont think we need to have a config for each LSP feature as long as they implementation does work.

**Comment by kubawolanin** [comment]
Hi Samuel,

> should we differentiate between local and remote lsp in config?

That really depends on what config properties do you plan to incorporate for the local and if they are cannot be applied for the remote.

> openhab.lspEnabled used for both or use two settings instead?

AFAIK we should keep the local language server online without a config flag turning it off. Most VSCode extensions operate that way. Thus, I would suggest renaming `lspEnabled` to `remoteLspEnabled` or a like.

> openhab.restCompletions this is now in the local LSP and does not slow down code completion anymore, so should we keep this or remove it?

We can remove it, thanks :)

**Comment by SamuelBrucksch** [comment]
Hm, looks like my rebase added some more commits than were necessary... i did 'git rebase origin-remote/master' was that wrong?

**Comment by SamuelBrucksch** [comment]
Should i add changelog entries already or is this done seperately? The current changelog would look something like this:
## 0.5.0 -TBD
- Added local LSP server (#122)
- Fixed sorting order in items explorer (#125)
- removed settings param 'restCompletions'
- renamed setting param 'lspEnabled' to 'remoteLspEnabled'
- renamed setting param 'lspPort' to 'remoteLspPort'

**Comment by MHerbst** [comment]
There seems to be a problem with outline view. With the test vsix (https://github.com/openhab/openhab-vscode/pull/122#issuecomment-449644967) I did not see any outline. After I went back to the marketplace version the outline was OK.

**Comment by SamuelBrucksch** [comment]
Outline works for me (checked in sitemap, rules, items). Seems like outline items are provided by remote LSP so make sure it is enabled to see the outline. At least when i deisabled remote lsp no outline was there, but when i enabled it, it was there...

**Comment by SamuelBrucksch** [comment]
Latest vsix:
[openhab-0.4.1.zip](https://github.com/openhab/openhab-vscode/files/2710896/openhab-0.4.1.zip)

**Comment by SamuelBrucksch** [comment]
Come on guys, this is important to work on other tasks as well.

**Comment by SamuelBrucksch** [comment]
Hi, about the items view: i noticed that as well. I think we can also use the completion items from LSP for that. Behaviour did not change, it was like that from the start so i wanted to do this in a seperate PR later. Not sure ifwe actually need the values in completions at all, i thought about a hover provider that displays the values on hover of an item in code.

**Comment by MHerbst** [comment]
> Not sure ifwe actually need the values in completions at all, i thought about a hover provider that displays the values on hover of an item in code.

In completions completion the item names and an information about the type are sufficient.
Hover with current value would be nice :-)

**Comment by Confectrician** [comment]
Hi all,

I finally had some time for (functional) testing and could not recognize any missbehavior for now.
I didn't look at the code yet in detail, but will do so at least for the ts parts.

Sorry for the delay, but we are all doing this in our free time besides work/school/study/whatever.
I know it can be annoying sometimes, but that's one of the compromises open source can bring.

**Comment by Confectrician** [comment]
@SamuelBrucksch maybe you can squash your commits and add a sign off now?

**Comment by Confectrician** [comment]
Also https://app.codacy.com/app/Confectrician/openhab-vscode/pullRequest?prid=2701365 shows some issues.
All of them are located in the js server part and most of them are about quoting style.
Maybe you could have a look at that.
But since it doesnt reduce functionality for now thats not a must have for me.

**Comment by SamuelBrucksch** [comment]
Hi, whats the best way to squash them? I saw rebase -i works or reset soft and then merge --squash, but i guess all of them would result in a git push -f at the end so all thesingle commits are gone, is that right?

Double quotes for JS does not make much sense as there are also other ways to create strings like 

```
const number = 5
console.log(`Number is ${number}`)
```
This does not work with double quotes. Also its easier to use quotings and so on in single quotes without disturbing escape chatacters.

**Comment by Confectrician** [comment]
>so all thesingle commits are gone

Yep but this doesn't matter in this state.
We are doing squash and merge in openhab repos anyway.
So at the latest when merging, this will be one commit in master.

We can leave it too, its just some cosmeticcal stuff to satisfy pullapprove . 🙂

**Comment by Confectrician** [comment]
@MHerbst or @kubawolanin any additional comments?
We can change stuff anyway and from my tests its functional, so i would say lgtm.

**Comment by Confectrician** [comment]
:shipit:

[Approved with PullApprove](https://pullapprove.com/openhab/openhab-vscode/pull-request/122/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=openhab/openhab-vscode)

File: serverJS/src/Server.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+  /*
+  documentClosed (event) {
+    // TODO
+  }
+  */
+
+  /*
+  documentSaved (event) {
+    // TODO can be used for format on save feature
```
[the remark this answers is not in the document]
Comment (Confectrician) [reply serverJS/src/Server.js]: I would remove them too.
Those are just empty functions.

We could open an issue instead, which reminds us that these functions (names written down) could be used later.

**Comment by SamuelBrucksch** [comment]
I removed the commented code. Do you still want me to squash everything even if is squashed anyways?

**Comment by kubawolanin** [comment]
> I removed the commented code. 

Perfect, thanks!

> Do you still want me to squash everything even if is squashed anyways?
I think we're good here, since you've signed off your first comment and committed all code with one git user :-)
Let's get this on master now!

Thank you for your work.

**Comment by SamuelBrucksch** [comment]
Yay! Thanks for reviewing. Now i will add a lot more PRs ;).

**Pull request merged**